### ISSUE TITLE
[flashlight-build] Error cross compiling from Linux 64bit

### ISSUE THREAD

**Issue by aranhoide** [getlantern/lantern#2258]
```
aranhoide@duna:~/src/flashlight-build$ ./crosscompile.bash
Generating UI resources for embedding
Dist folder is up to date
Generating resources.go
Now embedding lantern.ico to windows executable
Icon  lantern.ico  ID:  14
fatal: no tag exactly matches '254701747ca5babb740d6c477a7cfcfbb0516b94'
Building flashlight version 2547017 (20150223%.234808)
Number of parallel builds: 4

-->    darwin/amd64: github.com/getlantern/flashlight

1 errors occurred:
--> darwin/amd64 error: exit status 2
Stderr: # github.com/getlantern/systray
gcc: error: unrecognized command line option '-fobjc-arc'
```

**Comment by xiam** [comment]
Just had the same issue while playing with the docker image, and you're the only reference for it! I think you missed the CGO_ENABLED=1 env while bootstrapping Go:

```
CGO_ENABLED=1 GOOS=darwin GOARCH=amd64 ./make.bash
```

But you're using it when trying to compile flashlight.

Anyways, I haven't succeed at running ./make.bash with CGO_ENABLED=1 for darwin yet, at one point it says:

```
# net
could not determine kind of name for C.AI_MASK
```

Still trying.

**Comment by fffw** [comment]
@xiam What's your OS version inside docker image?

**Comment by atavism** [comment]
@fffw It's Ubuntu 14.04

**Comment by fffw** [comment]
As we are using gox in `crosscompile.sh`, is `gox -build-toolchain` enough?

`
gcc: error: unrecognized command line option '-fobjc-arc'
`
ARC only available in Apple LLVM Compiler, so gcc will report this error. Obj-C code should be changed to make it happy.

As we can already compile Darwin and Windows from Mac, can we just target Linux for x86 and arm inside docker?

**Comment by xiam** [comment]
Hello @fffw!

> As we can already compile Darwin and Windows from Mac, can we just target Linux for x86 and arm inside docker?

Yes, supporting linux/amd64 and linux/386 is our main target. I don't think it would be possible to cross compile Linux for OSX from within a Linux box, as long as we depend on C features for systray, but I also don't consider that a problem. 

I'm currently having some issues with linux/386, my docker build ends with this:

````
# github.com/getlantern/systray
/usr/bin/ld: cannot find -lappindicator3
/usr/bin/ld: cannot find -ldbusmenu-glib
/usr/bin/ld: cannot find -lgtk-3
/usr/bin/ld: cannot find -lgdk-3
/usr/bin/ld: cannot find -latk-1.0
/usr/bin/ld: cannot find -lgio-2.0
/usr/bin/ld: cannot find -lpangocairo-1.0
/usr/bin/ld: cannot find -lgdk_pixbuf-2.0
/usr/bin/ld: cannot find -lcairo-gobject
/usr/bin/ld: cannot find -lpango-1.0
/usr/bin/ld: cannot find -lcairo
/usr/bin/ld: cannot find -lgobject-2.0
/usr/bin/ld: cannot find -lglib-2.0
collect2: error: ld returned 1 exit status
```

All those libraries exist on the system as amd64 objects, but the i386 versions are missing, and I can't just install them without causing a big mess on the system...

**Comment by myleshorton** [comment]
I wonder if  linux/amd64 is enough and maybe we don't need linux/386. It
would be interesting to look at the download numbers for each.

On Tue, Mar 17, 2015 at 7:49 PM, Carlos Nieto <notifications@github.com>
wrote:

> Hello @fffw <https://github.com/fffw>!
>
> As we can already compile Darwin and Windows from Mac, can we just target
> Linux for x86 and arm inside docker?
>
> Yes, supporting linux/amd64 and linux/386 is our main target. I don't
> think it would be possible to cross compile Linux for OSX from within a
> Linux box, as long as we depend on C features for systray, but I also don't
> consider that a problem.
>
> I'm currently having some issues with linux/386, my docker build ends with
> this:
>
> # github.com/getlantern/systray
> /usr/bin/ld <http://github.com/getlantern/systray/usr/bin/ld>: cannot find -lappindicator3
> /usr/bin/ld: cannot find -ldbusmenu-glib
> /usr/bin/ld: cannot find -lgtk-3
> /usr/bin/ld: cannot find -lgdk-3
> /usr/bin/ld: cannot find -latk-1.0
> /usr/bin/ld: cannot find -lgio-2.0
> /usr/bin/ld: cannot find -lpangocairo-1.0
> /usr/bin/ld: cannot find -lgdk_pixbuf-2.0
> /usr/bin/ld: cannot find -lcairo-gobject
> /usr/bin/ld: cannot find -lpango-1.0
> /usr/bin/ld: cannot find -lcairo
> /usr/bin/ld: cannot find -lgobject-2.0
> /usr/bin/ld: cannot find -lglib-2.0
> collect2: error: ld returned 1 exit status
>
> All those libraries exist on the system as amd64 objects, but the i386
> versions are missing, and I can't just install them without causing a big
> mess on the system...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/2258#issuecomment-82698192>.
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by oxtoacart** [comment]
I don't know about downloads, but in terms of visitors to our site, for Linux, 77.94% of visitors in the last month were using x86_64 (64-bit) and 18.82% were using i686 (32 bit).

We may need to think about ARM to because of Lantern PI.

**Comment by xiam** [comment]
I ditched Ubuntu as the base of the image and tried Fedora instead, now I can tell that Fedora has better support for multiarch! This image is able to compile Lantern binaries for linux/amd64, linux/386 and windows/386 from any docker-capable host.

I moved the Dockerfile repo to here: https://github.com/getlantern/lantern-docker-builder

I think this docker image can also be used to generate .deb packages, I didn't continue with that since we are not going to generate packages that often, if you see the source I commented out the lines that would install the required packages that can be used to generate .deb packages.

I'll get back to generating full packages later, I have a pending investigation on the mobile scene (with @atavism).

**Comment by xiam** [comment]
I just finished my testing on https://github.com/getlantern/lantern-docker-builder, the binary created with the Fedora docker image on a OSX host runs perfectly on Ubuntu.

@myleshorton You had an issue with `genassets.sh`, it turned out to be more easy to fix than I thought, that was causing your last test to fail, I added an example on how to build a fully working executable too.

So maybe there's no need to ditch support 386 for now :-).

(Should we close this yet?)
